File: array-easy/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
class Solution:
    # Solution 1: O(n2)
    #     def twoSum(self, nums: List[int], target: int) -> List[int]:
    #         count = len(nums)
    #         for i in range(count):
    #             num = nums[i]

    #             for j in range(i + 1, count):
    #                 if (num + nums[j] == target):
    #                     return [i, j]
    #         return [0, 0]

    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hash = {}

        for i in range(len(nums)):
            num = nums[i]
            complement = target - num

            if num in hash:
                return [hash[num], i]

            hash[complement] = i

        return [0, 0]


# @lc code=end

File: array-easy/1.two-sum.cs
/*
 * @lc app=leetcode id=1 lang=csharp
 *
 * [1] Two Sum
 */

// @lc code=start
public class Solution
{
    public int[] TwoSum(int[] nums, int target)
    {
        var len = nums.Length;

        // Solution 1: O(n2), O(1)
        /*
        for (var i = 0; i < len - 1; i++) {
            for (var j = i + 1; j < len; j++) {
                if (nums[i] + nums[j] == target) {
                    return new int[2] {i, j};
                }
            }
        }
        
        return null;
        */

        // Solution 2: O(n), O(n)
        var hash = new Dictionary<int, int>();
        var num = 0;
        for (var i = 0; i < len; i++)
        {
            num = nums[i];
            var complement = target - num;
            if (hash.ContainsKey(complement))
            {
                return new int[] { hash[complement], i };
            }

            if (!hash.ContainsKey(num))
            {
                hash.Add(nums[i], i);
            }
        }

        return null;
    }
}
// @lc code=end

